align/schema/types.py

```diff
@@ -41,11 +47,39 @@ def __delitem__(self, sliceobj):
     def __eq__(self, other):
         return self.__root__ == other
 
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._commits = collections.OrderedDict()
+
+    def _gen_commit_id(self, nchar=8):
+        id_ = ''.join(random.choices(string.ascii_uppercase + string.digits, k=nchar))
+        return self._gen_commit_id(nchar) if id_ in self._commits else id_
+
+    def checkpoint(self):
+        self._commits[self._gen_commit_id()] = len(self)
+        return next(reversed(self._commits))
+
+    def _revert(self):
+        _, length = self._commits.popitem()
+        del self[length:]
+
+    def revert(self, name=None):
+        assert len(self._commits) > 0, 'Top of scope. Nothing to revert'
+        if name is None or name == next(reversed(self._commits)):
+            self._revert()
+        else:
+            assert name in self._commits
+            self._revert()
+            self.revert(name)
+
 class Dict(pydantic.generics.GenericModel, typing.Generic[KeyT,DataT]):
```

Comment:
1. For educational purpose: what is the benefit of defining/customizing `Dict` or `List`? How are these different than the default ones?
2. Does `Dict` object allow: `if key in instance_Dict:` ? This was throwing an error for me.

Reply:
We needed to inherit from pydantic.BaseModel to allow `copy_on_model_validation=False` to work for container objects as well. I also added a checkpoint / revert type of mechanism to this custom List that may prove useful anywhere we need incremental upgrades to the same list (Currently used for ConstraintDB).

the `in` operator is probably not working because I forgot to implement `__contains__`